libmamba/include/mamba/core/util_string.hpp

```diff
@@ -112,86 +114,146 @@ namespace mamba
                 left += right;
             }
         };
-    }
 
-    template <class S, class CharType, class Joiner = details::PlusEqual>
-    auto join(const CharType* sep, const S& container, Joiner joiner = details::PlusEqual{}) ->
-        typename S::value_type
-    {
-        if (container.empty())
-            return {};
-        auto result = container[0];
-        for (std::size_t i = 1; i < container.size(); ++i)
+        template <class T, class = void>
+        struct has_reserve : std::false_type
+        {
+        };
+        template <class T>
+        struct has_reserve<T, std::void_t<decltype(std::declval<T>().reserve(std::size_t()))>>
+            : std::true_type
+        {
+        };
+        template <typename T>
+        inline constexpr bool has_reserve_v = has_reserve<T>::value;
+
+        std::size_t size(const char* s);
```

Comment:
Not sure why these are necessary, but you can move the definition here as inline, it's literally only calls to the standard library which will be inlined and using the cpp for that prevents the inlining.

Reply:
It's for generic code below (*e.g.* in `concat` to call it over a parameter pack that may contain, `std::string`s, `char`s, `const char*`...).
It's a modification of what was already there.